### PR TITLE
One-pass computation of line spacing

### LINKED ISSUE #1633: Collisions with text above and below staff
Presumably both of these have similar fixes, but I haven't looked into it:

```
\documentclass{article}
\usepackage{gregoriotex}
\begin{document}

\gresetinitiallines{0}

{\gresetlineheightexpansion{uniform}
  \gabcsnippet{(c3) <alt>abovelines</alt>g(g) c4(c4)}
}

\gabcsnippet{(c2) gz0::c4(gz0::c4) e(e)}

\end{document}
```

Image

### PR DESCRIPTION
**Pull request by davidweichiang**
In #1488 the question came up of whether variable line spacing could be done in a single pass. This (sorry, rather large) patch does that. The score lines are printed in TeX with possible collisions, but with different parts of the score distinguished by different attributes, so that they can be put into the proper place in Lua. The advantages of this I see are:

- With two passes, the user might forget to run a second time, or be confused by large changes between the first and second pass.
- The two-pass computation was complicated, with control passing from TeX to Lua to TeX to Lua. It was difficult to inject the `\grechangedim`s at exactly the end of a line, which led to bugs.

The disadvantages are:

- `\grechangenextscorelinedim` and `\grechangenextscorelinecount` now only work for the distances and counts used for line spacing. I guess a helpful error should be printed if the user tries to change something else.
- It might seem conceptually strange to print a bad-looking score at first (which no one would see). In particular, the translation is printed right on top of the lyrics.

All tests pass except for:
gabc-output/glyphs/clef_change.gabc : this is #1633 I think
gabc-output/glyphs/nabc-edge-cases.gabc : this is #1634
gabc-output/bugs/fix-569.gabc : this is #1633
gabc-output/lines/3-lines.gabc : this is #1636
gabc-output/lines/2-lines.gabc : #1636 again
gabc-output/lines/5-lines.gabc : #1636 again
tex-output/dominican-flats/dominican-flats.tex : this one fails for me even in `ctan`, and the "correct" output looks wrong; I must have messed it up somehow
tex-output/bar-spacing/bar-spacing-new.tex : the difference is only in the lines used for debugging, and the new lines look more consistent to me?

Supersedes #1639.
Closes #1633, #1634, #1636.

### CODE REVIEW

**Comment by rpspringuel** [comment]
On the tests:

gabc-output/glyphs/clef_change.gabc : The new results look correct, while the old ones have a bad collision.
gabc-output/glyphs/nabc-edge-cases.gabc : I can confirm this fails, but the new results look fine to me.
gabc-output/bugs/fix-569.gabc : Again, new results seem to be better.
gabc-output/lines/3-lines.gabc : Definately a fix for the custos position.
gabc-output/lines/2-lines.gabc : Same as for 3-lines
gabc-output/lines/5-lines.gabc : Same as for 3-lines
tex-output/dominican-flats/dominican-flats.tex : This one doesn't fail for me.
tex-output/bar-spacing/bar-spacing-new.tex : I agree, the new results look like they are placing the debugging lines better.

tex-output/hidelyricsnotes/text-only.tex fails for me.  The document doesn't compile and the log indicates theirs a problem trying to do math that probably shouldn't be done in this case: `gregoriotex.lua:744: attempt to perform arithmetic on a nil value (field 'glyph_top')`

Some comments:
>With two passes, the user might forget to run a second time, or be confused by large changes between the first and second pass.

Multiple passes are a standard part of preparing TeX documents, so this would only be a factor for those who are new to the ecosystem.

That said, getting rid of requiring the second pass ourselves could be a huge time savings.  Loading GregorioTeX takes a substantial amount of time so while we probably can't get rid of the need for a second pass all by ourselves, if we can make it so that we are requiring the second pass, then we at least make the single-pass documents using Gregorio possible.  If we can solve some bugs at the same time, all the better.

> `\grechangenextscorelinedim` and `\grechangenextscorelinecount` now only work for the distances and counts used for line spacing. I guess a helpful error should be printed if the user tries to change something else.

The documentation indicates that `\grechangenextscorelinedim`is only supposed to work with `spaceabovelines`, `spacebeneathtext`, and `spacelinestext`.  That it worked with other distances is therefore, an undocumented feature, so your "restriction" would actually be an expansion of what it is documented to work with.

Besides, in both cases, I think this is what we originally intended both of these macros to be for.

>It might seem conceptually strange to print a bad-looking score at first (which no one would see). In particular, the translation is printed right on top of the lyrics.

This is hardly any more strange than the intermediate document produced in the middle of a multi-pass system.

**Comment by rpspringuel** [comment]
Oh, and if you could create a parallel branch in the test repository with the new test results that would be appreciated (at least for the tests where we know the new result is because we're fixing the bugs).

**Comment by davidweichiang** [comment]
> tex-output/dominican-flats/dominican-flats.tex : This one doesn't fail for me.

It must be because I'm working off of https://github.com/gregorio-project/gregorio-test/pull/360 and it's gotten out of sync.

> tex-output/hidelyricsnotes/text-only.tex fails for me. The document doesn't compile and the log indicates theirs a problem trying to do math that probably shouldn't be done in this case: gregoriotex.lua:744: attempt to perform arithmetic on a nil value (field 'glyph_top')

My fault, that one fails for me too, but it should work now with the last commit.

**Comment by davidweichiang** [comment]
There's still a command `\greskipheightcomputation` that still does something, but not what it says. What should it be called instead?

**Comment by davidweichiang** [comment]
Since the speed difference due to `\greskipheightcomputation` seemed to be negligible (<0.1%), I took the liberty of removing it.

**Comment by rpspringuel** [comment]
I was away for the weekend and have been sick since I got back, so I haven’t had a chance to review anything posted since last Wednesday. I’m hoping to feel better and get back to work by Monday. 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by davidweichiang** [comment]
With the last commit, three additional tests fail:
tex-output/commentary/commentary.tex: because the commentary height takes note height into account (#1474)
tex-output/PopulusSion/PopulusSion.tex: because a small bug was causing the commentary to be printed loo low
tex-output/per-line-spaces/per-line-spaces.tex: same as previous

**Comment by davidweichiang** [comment]
> I was away for the weekend and have been sick since I got back, so I haven’t had a chance to review anything posted since last Wednesday. I’m hoping to feel better and get back to work by Monday.

No worries and hope you feel better soon!

File: tex/gregoriotex-spaces.tex [hunk 1]
```diff
@@ -1590,6 +1590,9 @@
 % #1 : line number
 % #2-#4 : same as #1-#3 of \grechangedim
 \def\grechangenextscorelinedim#1#2#3#4{%
+  \IfStrEqCase{#2}{{abovelinestextheight}{}{noteadditionalspacelinestext}{}{spaceabovelines}{}{abovelinestextraise}{}{spacelinestext}{}{spacebeneathtext}{}{translationheight}{}}[%
```
Comment (rpspringuel) [inline tex/gregoriotex-spaces.tex:1590]: Just to make sure I'm reading this right, you're using `\IfStrEqCase` to construct a `not (... and ... and ...)` sequence that will raise an error if this macro is called with an unsupported distance.

File: tex/gregoriotex-main.tex [hunk 2]
```diff
@@ -1117,10 +1067,10 @@
 
 %macro called at the beginning of a score
 % #1 is the gabc score id
-% #2 is the high height
-% #3 is the low height
-% #4 is 1 if there is a translation somewhere
-% #5 is if 1 if we have space above the staff
+% #2 is the high height (currently ignored)
```
Comment (rpspringuel) [inline tex/gregoriotex-main.tex]: Can you foresee a future use for these arguments?  If not, I'd mark them as deprecated, not just ignored.  Once this PR is approved and merged, then we can work on the appropriate PR for the develop branch which will actually remove them (since they'd have to be removed from the executable, that's not a ctan eligible change).

**Comment by rpspringuel** [comment]
I agree that the newly broken tests are all fixes to something which is currently wrong.

File: tex/gregoriotex-spaces.tex (hunk 1 above)
Comment (davidweichiang) [reply tex/gregoriotex-spaces.tex:1590]: That's right.

Comment (davidweichiang) [reply tex/gregoriotex-spaces.tex:1590]: Oh, actually, it's `if not (... or ... or ...) then print error`.

File: tex/gregoriotex-main.tex (hunk 2 above)
Comment (davidweichiang) [reply tex/gregoriotex-main.tex]: I don't foresee a future use, no. How do I mark them as deprecated?

Comment (rpspringuel) [reply tex/gregoriotex-main.tex]: I've updated the Wiki page on the Deprecation process to cover this case: https://github.com/gregorio-project/gregorio/wiki/Deprecation-Process

File: tex/gregoriotex-spaces.tex (hunk 1 above)
Comment (rpspringuel) [reply tex/gregoriotex-spaces.tex:1590]: Yeah, that's what I meant.

**Comment by davidweichiang** [comment]
Done, I think

File: src/gregoriotex/gregoriotex-write.c [hunk 3]
```diff
@@ -4701,9 +4701,11 @@ void gregoriotex_write_score(FILE *const f, gregorio_score *const score,
     }
     fprintf(f, "\\GreBeginScore{%s}{%d}{%d}{%d}{%d}{%s}{%u}"
```
Comment (rpspringuel) [inline src/gregoriotex/gregoriotex-write.c:4707]: I'd put a deprecation comment here indicating what the string of arguments will look like after the deprecated ones are removed.

Comment (davidweichiang) [reply src/gregoriotex/gregoriotex-write.c:4707]: OK (also I wrote comments on two struct fields that can be removed).

**Comment by davidweichiang** [comment]
I won't tinker with this anymore -- hopefully it's ready to go?

**Comment by rpspringuel** [comment]
In reviewing this I've come across two unexplained test failures:

gabc-output/queuelength2_legacy.gabc
gabc-output/queuelength2.gabc

In both cases only the last three staves of the second page are effected.  The third to last stave is pushed down (though its lyrics are pushed up) and the last two staves are pushed up (by the same amount as the lyrics in the third to last stave, from the looks of it).

All the other test changes look good and are taken care of by the PR https://github.com/gregorio-project/gregorio-test/pull/373

**Comment by davidweichiang** [comment]
Thanks for catching this -- I promised I wouldn't tinker any more and then I did! I believe this is a bug fix.

The first note of this score is below the staff (c=3), and I didn't fully look into why, but in `ctan` and in all commits of this PR up to the last one, the attribute `\gre@attr@glyph@bottom`=3 reappears in various places throughout the whole score. Ordinarily this seems to be harmless, but on line 14 (third from last), and again I didn't fully look into why, the 3 actually appears as an attribute on an hbox that matters. The line spacing algorithm thinks this line has a `c` note on it, and the lyrics are typeset too low.

image

I believe the shift in the last two staves is because in the old code, line 14's height was greater than `\baselineskip`, so `\lineskip` (=1pt, I think) was inserted.

In the last commit, `\gre@attr@glyph@top` and `\gre@attr@glyph@bottom` are reset after each glyph, so the problem goes away.

image

**Comment by rpspringuel** [comment]
That explains why the lyrics (and the staves below) move up, but why does the 3rd to last set of staff lines move down?  There doesn't appear to be anything on the 4th to last line that would require extra space below that stave nor anything on the 3rd to last line that would require extra space above that line.

**Comment by davidweichiang** [comment]
That's because the baseline of each staff is the baseline of the lyrics (or translation if any), and TeX tries to ensure that the distance between baselines is `\baselineskip`. So lyrics moving up really means staff moving down.

(I think a case could be made for the baseline to instead be the lowest staff line, but that's a separate issue!)

**Comment by rpspringuel** [comment]
This is why I find `\baselineskip` so confusing.

Okay, so it looks like everything this good at this point, just need to have those altered tests accepted into the test repository.

**Comment by davidweichiang** [comment]
Done, thanks!
